File: backend/models/media_studio.go
package models

import pb "goodanime-backend/proto"

// ToProto converts Studio to its protobuf representation.
func (s *Studio) ToProto() *pb.Studio {
	if s == nil {
		return nil
	}

	return &pb.Studio{
		Id:          s.ID,
		Name:        s.Name,
		IsFavourite: s.IsFavourite,
		Favourites:  s.Favourites,
		Media:       s.Media.ToProto(),
	}
}

// ToProto converts StudioMediaConnection to its protobuf representation.
func (sm *StudioMediaConnection) ToProto() *pb.StudioMediaConnection {
	if sm == nil {
		return nil
	}
	res := &pb.StudioMediaConnection{
		PageInfo: sm.PageInfo.ToProto(),
	}
	for _, node := range sm.Nodes {
		res.Nodes = append(res.Nodes, node.ToProto())
	}
	return res
}

// ToProto converts Studio to its protobuf representation.
func (s *StudioMin) ToProto() *pb.StudioMin {
	if s == nil {
		return nil
	}
	return &pb.StudioMin{
		Id:   s.ID,
		Name: s.Name,
	}
}

// ToProto converts StudioEdge to its protobuf representation.
func (e *StudioEdge) ToProto() *pb.StudioEdge {
	if e == nil {
		return nil
	}
	return &pb.StudioEdge{
		IsMain: e.IsMain,
		Node:   e.Node.ToProto(),
	}
}

// ToProto converts StudioConnection to its protobuf representation.
func (c *StudioConnection) ToProto() *pb.StudioConnection {
	if c == nil {
		return nil
	}
	res := &pb.StudioConnection{}
	for _, edge := range c.Edges {
		res.Edges = append(res.Edges, edge.ToProto())
	}
	return res
}

// StudioMin represents a minimal production studio
type StudioMin struct {
	ID   int32  `json:"id"`
	Name string `json:"name"`
}

// StudioEdge represents a link between a media and a studio
type StudioEdge struct {
	IsMain bool      `json:"isMain"`
	Node   StudioMin `json:"node"`
}

// StudioConnection represents a list of studio edges
type StudioConnection struct {
	Edges []StudioEdge `json:"edges"`
}

// Studio represents a production studio
type Studio struct {
	ID          int32                  `json:"id"`
	Name        string                 `json:"name"`
	Favourites  *int32                 `json:"favourites"`
	IsFavourite *bool                  `json:"isFavourite"`
	Media       *StudioMediaConnection `json:"media"`
}

// StudioMediaConnection represents a list of media associated with a studio
type StudioMediaConnection struct {
	PageInfo PageInfo   `json:"pageInfo"`
	Nodes    []MediaMin `json:"nodes"`
}

type StudioDTO struct {
	Studio Studio `json:"Studio"`
}

type StudioSearchDTO struct {
	Page struct {
		PageInfo PageInfo  `json:"pageInfo"`
		Studios  []*Studio `json:"studios"`
	} `json:"Page"`
}

File: lib/models/media_studio.dart
import '../proto/media_min.pb.dart' as pbm;
import '../proto/media_studio.pb.dart' as pb;
import 'common.dart';
import 'media_min.dart';

/// Represents a minimal production studio
class StudioMin {
  /// Studio ID
  final int id;

  /// Studio name
  final String name;

  /// Creates a StudioMin
  const StudioMin({required this.id, required this.name});

  /// Creates a StudioMin from a protobuf object
  factory StudioMin.fromProto(pb.StudioMin pbObj) {
    return StudioMin(id: pbObj.id, name: pbObj.name);
  }

  /// Converts the studio to a protobuf object
  pb.StudioMin toProto() {
    return pb.StudioMin(id: id, name: name);
  }
}

/// Represents a link between a media and a studio
class StudioEdge {
  /// Whether this is the main studio
  final bool isMain;

  /// The studio
  final StudioMin node;

  /// Creates a studio edge
  const StudioEdge({required this.isMain, required this.node});

  /// Creates a studio edge from a protobuf object
  factory StudioEdge.fromProto(pb.StudioEdge pbObj) {
    return StudioEdge(
      isMain: pbObj.isMain,
      node: StudioMin.fromProto(pbObj.node),
    );
  }

  /// Converts the studio edge to a protobuf object
  pb.StudioEdge toProto() {
    return pb.StudioEdge(isMain: isMain, node: node.toProto());
  }
}

/// Represents a studio
class Studio {
  /// The studio's ID
  final int id;

  /// The studio's name
  final String name;

  /// The number of favourites the studio has
  final int? favourites;

  /// Whether the studio is a favourite of the user
  final bool? isFavourite;

  /// The studio's media connection
  final StudioMediaConnection? media;

  /// Creates a studio
  const Studio({
    required this.id,
    required this.name,
    required this.favourites,
    required this.isFavourite,
    required this.media,
  });

  /// Creates a studio from a protobuf object
  factory Studio.fromProto(pb.Studio pbObj) {
    return Studio(
      id: pbObj.id,
      name: pbObj.name,
      favourites: pbObj.hasFavourites() ? pbObj.favourites : null,
      isFavourite: pbObj.hasIsFavourite() ? pbObj.isFavourite : null,
      media: pbObj.hasMedia()
          ? StudioMediaConnection.fromProto(pbObj.media)
          : null,
    );
  }

  /// Converts the studio to a protobuf object
  pb.Studio toProto() {
    return pb.Studio(
      id: id,
      name: name,
      favourites: favourites,
      isFavourite: isFavourite,
      media: media?.toProto(),
    );
  }
}

/// Represents a studio's media connection
class StudioMediaConnection {
  /// The page info for the media connection
  final PageInfo pageInfo;

  /// The list of media nodes
  final List<MediaMin> nodes;

  /// Creates a studio media connection
  const StudioMediaConnection({required this.pageInfo, required this.nodes});

  /// Creates a studio media connection from a protobuf object
  factory StudioMediaConnection.fromProto(pb.StudioMediaConnection pbObj) {
    return StudioMediaConnection(
      pageInfo: PageInfo.fromProto(pbObj.pageInfo),
      nodes: pbObj.nodes.map(MediaMin.fromProto).toList(),
    );
  }

  /// Converts the studio media connection to a protobuf object
  pb.StudioMediaConnection toProto() {
    return pb.StudioMediaConnection(
      pageInfo: pageInfo.toProto(),
      nodes: nodes.map((node) => node.toProto() as pbm.MediaMin).toList(),
    );
  }
}
